tensorflow_decision_forests/contrib/temporal_feature_processor/temporal_feature_processor/core/evaluator.py

```diff
@@ -14,24 +14,98 @@
-
-Result = Data
+AvailableBackends = Literal[tuple(backends.BACKENDS.keys())]
+Data = Dict[Event, Union[str, pathlib.Path, pandas_event.PandasEvent]]
+Query = Union[Event, List[Event]]
 
 
-def Eval(
+def evaluate(
     query: Query,
-    data: Data,
-) -> Result:
+    input_data: Data,
+    backend: AvailableBackends = "pandas",
+) -> Dict[Event, Any]:
   """Evaluates a query on data."""
 
-  del query
-  del data
+  if isinstance(query, Event):
+    events_to_compute = [query]
+
+  elif isinstance(query, list):
+    events_to_compute = query
+
+  else:
+    # TODO: improve error message
+    raise TypeError(
+        f"schedule_graph query argument must be one of {Query}. Received {type(query)}."
+    )
+
+  # get features from events
+  features_to_compute = [
+      feature for event in events_to_compute for feature in event.features()
+  ]
+
+  # get backend
+  selected_backend = backends.BACKENDS[backend]
+  event = selected_backend["event"]
+  evaluate_schedule_fn = selected_backend["evaluate_schedule_fn"]
+  read_csv_fn = selected_backend["read_csv_fn"]
+
+  # calculate opeartor schedule
+  schedule = get_operator_schedule(features_to_compute)
+
+  # materialize input data. TODO: separate this logic
+  materialized_input_data = {
+      input_event:
+      input_event_spec if isinstance(input_event_spec, event) else read_csv_fn(
+          input_event_spec, input_event.sampling())
+      for input_event, input_event_spec in input_data.items()
+  }
+  # evaluate schedule
+  outputs = evaluate_schedule_fn(materialized_input_data, schedule)
+
+  return {event: outputs[event] for event in events_to_compute}
+
+
+def get_operator_schedule(query: Set[Feature]) -> List[base.Operator]:
+  # TODO: add depth calculation for parallelization
+
+  operators_to_compute = []  # TODO: implement as ordered set
+  visited_features = set()
+  pending_features = list(query.copy())  # TODO: implement as ordered set
+  while pending_features:
+    feature = next(iter(pending_features))
+    visited_features.add(feature)
+
+    if feature.creator() is None:
+      # is input feature
+      pending_features.remove(feature)
+      continue
+
+    # required input features to compute this feature
+    creator_input_features = {
+        input_feature for input_event in feature.creator().inputs().values()
+        for input_feature in input_event.features()
+    }
+    # check if all required input features have already been visited
+    if creator_input_features.issubset(visited_features):
+      # feature can be computed - remove it from pending_features
+      pending_features.remove(feature)
+
+      # add operator at the end of operators_to_compute
+      if feature.creator() not in operators_to_compute:
+        operators_to_compute.append(feature.creator())
+
+      continue
+
+    # add required input features at the beginning of pending_features
+    pending_features = list(creator_input_features) + pending_features
```

Comment:
If a feature is used as input by multiple ops (i.e. the feature is already in the pending_features), could this create unnecessary duplicates? While not optimal, why not simply removing the elements "creator_input_features" from "pending_features" before the addition?

Reply:
Good catch! I'll implement this right away.